monai/apps/detection/utils/anchor_utils.py

```diff
@@ -0,0 +1,408 @@
+
+        ws = (w_ratios[:, None] * scales_t[None, :]).view(-1)
+        hs = (h_ratios[:, None] * scales_t[None, :]).view(-1)
+        if self.spatial_dims == 2:
+            base_anchors = torch.stack([-ws, -hs, ws, hs], dim=1) / 2.0
+        elif self.spatial_dims == 3:
+            ds = (d_ratios[:, None] * scales_t[None, :]).view(-1)
+            base_anchors = torch.stack([-ws, -hs, -ds, ws, hs, ds], dim=1) / 2.0
+
+        return base_anchors.round()
+
+    def set_cell_anchors(self, dtype: torch.dtype, device: torch.device):
+        """
+        Convert each element in self.cell_anchors to ``dtype`` and send to ``device``.
+        """
+        self.cell_anchors = [cell_anchor.to(dtype=dtype, device=device) for cell_anchor in self.cell_anchors]
+
+    def num_anchors_per_location(self):
+        """
+        Return number of anchor shapes for each feature map.
+        """
+        return [c.shape[0] for c in self.cell_anchors]
+
+    def grid_anchors(self, grid_sizes: List[List[int]], strides: List[List[Tensor]]) -> List[Tensor]:
+        """
+        Every combination of (a, (g, s), i) in (self.cell_anchors, zip(grid_sizes, strides), 0:spatial_dims)
+        corresponds to a feature map.
+        It outputs g[i] anchors that are s[i] distance apart in direction i, with the same dimensions as a.
+
+        Args:
+            grid_sizes: spatial size of the feature maps
+            strides: strides of the feature maps regarding to the original image
+
+        Example:
+            .. code-block:: python
+
+                grid_sizes = [[100,100],[50,50]]
+                strides = [[torch.tensor(2),torch.tensor(2)], [torch.tensor(4),torch.tensor(4)]]
+        """
+        anchors = []
+        cell_anchors = self.cell_anchors
+        assert cell_anchors is not None
+
+        if not (len(grid_sizes) == len(strides) == len(cell_anchors)):
+            raise ValueError(
+                "Anchors should be Tuple[Tuple[int]] because each feature "
+                "map could potentially have different sizes and aspect ratios. "
+                "There needs to be a match between the number of "
+                "feature maps passed and the number of sizes / aspect ratios specified."
+            )
+
+        for size, stride, base_anchors in zip(grid_sizes, strides, cell_anchors):
+            # for each feature map
+            device = base_anchors.device
+
+            # compute anchor centers regarding to the image.
+            # shifts_centers is [x_center, y_center] or [x_center, y_center, z_center]
+            shifts_centers = [
+                torch.arange(0, size[axis], dtype=torch.int32, device=device) * stride[axis]
+                for axis in range(self.spatial_dims)
+            ]
+
+            # to support torchscript, cannot directly use torch.meshgrid(shifts_centers).
+            shifts_centers = list(torch.meshgrid(shifts_centers[: self.spatial_dims], indexing="ij"))
```

Comment:
Should this be refactored to the monai utility function?

Reply:
I spent some time on this one today, `monai.networks.utils.meshgrid_ij` is not compatible with torchscript...

Reply:
Yes, I also spent some time, and found out to support torchscript, you have to spell it out like `torch.meshgrid([shifts_centers[0], shifts_centers[1]], indexing="ij")` instead of torch.meshgrid(shifts_centers, indexing="ij")

Reply:
removing the `indexing="ij"` will currently have the best compatibility across various versions. but it'll trigger a warning with the latest pytorch versions `torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument.`